### ISSUE TITLE
Question mark at the beginning of text

### ISSUE THREAD

**Issue by vafada** [lanceewing/agile#19]
kq2, when reading the sign at the tree

![bugtext](https://user-images.githubusercontent.com/285941/197420998-81ddfc11-514b-4c3a-ba21-dc8549bd82ef.PNG)

For reference where the bug happens:

https://youtu.be/_jjUaeG9YTw?t=131

see discussion in https://github.com/lanceewing/agile/pull/16

**Comment by vafada** [comment]
playing around with C# console app...

this seems to work:

```
using System.Text;

Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);


Encoding cp437 = Encoding.GetEncoding(437);

Console.WriteLine("string = " + cp437.GetString(new byte[] { 75, 81, 255, 50 }));
```

prints `KQ 2`

**Comment by vafada** [comment]
https://int10h.org/oldschool-pc-fonts/fontlist/font?ibm_ega_8x8

**Comment by lanceewing** [comment]
Yeah, that looks like the same charset as what WinAGI has in its "AGI Font" Help page:

![image](https://user-images.githubusercontent.com/8688804/197671055-7631d5d7-cf6f-4137-ac99-f3cfe19364d8.png)

The bitmap data in the IBM_BIOS_FONT constant inside the TextGraphics class should hopefully include all that.

I have a few hours free at the moment, so I'll try changing the AGILibrary to fix this.

**Comment by lanceewing** [comment]
Cool, this seems to work. The AGILibrary is storing message text as C# strings, which is in Unicode format. The code you specified converts the CP437 characters into the Unicode equivalent. All CP437 characters have a Unicode equivalent:

https://en.wikipedia.org/wiki/Code_page_437#Character_set

I also had to use the CP437 encoding when converting back from the string to bytes, when the TextGraphics glass renders the individual characters, and having done that, then the non-breaking space now shows as a space.

I will create a test text screen with all extended characters to confirm that this works for all of them, and if so, will then push this change.

**Comment by lanceewing** [comment]
Looks like everything is working for the extended char set:

![image](https://user-images.githubusercontent.com/8688804/197682139-192f325f-c6f9-41a4-8c4d-8f0252cd4c72.png)

What I did was edit the KQ2 sign text to include all the extended characters on the first 8 lines.

**Comment by lanceewing** [comment]
Unfortunately it is not working properly for some of the characters in the first 128 characters, such as some of the control codes:

![image](https://user-images.githubusercontent.com/8688804/197683722-c3844fd3-053e-42f9-b57f-7a7b0c2747dc.png)

It appears that the conversion is not consistent in both directions, i.e. for some characters, converting from CP437 to Unicode then back to CP437 will not result in the original byte. So I think there is still a bit more work to do on this.

I think we have the original issues from the KQ2 sign, and in fact all extended characters, sorted though.

I did think that we could perhaps store messages in the original byte array format, for the purposes of AGILE, as there isn't actually a reason to have them in string format. I would need to look at how big a change that is. I might look first at whether there is some way to support the characters that are not working properly with the current mechanism, but if I end up exhausting that, then storing the original message byte data might be the way to go.

**Comment by lanceewing** [comment]
Actually, AGI itself is unable to display some characters. I've loaded my modified KQ2 version into the original AGI interpreter running in Dosbox and see something for the first row of standard characters that isn't that different from AGILE. The original interpreter also prints x00 in the first three places, which is for character 0. Character 1 is then shown, i.e. the first face, after that. So that bit is correct.

AGILE is rendering a glyph for character 8, whereas AGI does not, in fact AGI also doesn't render character 7 on my test page because it is interpreting the character 8 as a backspace and so the character before it is removed, i.e. it actually moves the cursor back and removes the character. AGILE renders both, i.e. doesn't interpreter character 8 as a backspace. 

WinAGI claims that characters 8, 9, 10 and 13 do not render as glyphs. This isn't true in the case of 9. That one is rendered in the original AGI interpreter, but it is true of 8, 10, and 13. As mentioned above, 8 is interpreted as a backspace. In addition to this, 10 is interpreted as a linefeed and 13 as a carriage return. AGILE doesn't interpret them the same but instead renders them all as glyphs. It doesn't have any kind of text mode console logic.

So, in summary, backspace, linefeed and carriage return characters need some work. I think I will merge the current changes later today anyway, as it is already an improvement. I suspect that backspace, linefeed and carriage return might be handled differently in text vs graphics modes in AGI but will need to investigate.

**Comment by lanceewing** [comment]
I have now merged the changes to use CP 437 for the decoding of LOGIC messages. This fixes the original issue.
